state/remote/swift.go

```diff
@@ -32,50 +50,201 @@ func swiftFactory(conf map[string]string) (Client, error) {
 }
 
 func (c *SwiftClient) validateConfig(conf map[string]string) (err error) {
-	if val := os.Getenv("OS_AUTH_URL"); val == "" {
-		return fmt.Errorf("missing OS_AUTH_URL environment variable")
+	authUrl, ok := conf["auth_url"]
+	if !ok {
+		authUrl = os.Getenv("OS_AUTH_URL")
+		if authUrl == "" {
+			return fmt.Errorf("missing 'auth_url' configuration or OS_AUTH_URL environment variable")
+		}
+	}
+	c.authurl = authUrl
+
+	username, ok := conf["user_name"]
+	if !ok {
+		username = os.Getenv("OS_USERNAME")
+	}
+	c.username = username
+
+	userID, ok := conf["user_id"]
+	if !ok {
+		userID = os.Getenv("OS_USER_ID")
+	}
+	c.userid = userID
+
+	password, ok := conf["password"]
+	if !ok {
+		password = os.Getenv("OS_PASSWORD")
+		if password == "" {
+			return fmt.Errorf("missing 'password' configuration or OS_PASSWORD environment variable")
+		}
 	}
-	if val := os.Getenv("OS_USERNAME"); val == "" {
-		return fmt.Errorf("missing OS_USERNAME environment variable")
+	c.password = password
+
+	region, ok := conf["region_name"]
+	if !ok {
+		region = os.Getenv("OS_REGION_NAME")
+		if region == "" {
+			return fmt.Errorf("missing 'region_name' configuration or OS_REGION_NAME environment variable")
+		}
+	}
+	c.region = region
+
+	tenantID, ok := conf["tenant_id"]
+	if !ok {
+		tenantID = os.Getenv("OS_TENANT_ID")
+		if tenantID == "" {
+			tenantID = os.Getenv("OS_PROJECT_ID")
+		}
+	}
```

Comment:
Check out [MultiEnvDefaultFunc](https://github.com/hashicorp/terraform/blob/master/helper/schema/schema.go#L195). It might make sense to implement a modified local function in `swift.go` to handle cases where multiple environment variables can be used. Something like:

```go
func multiEnv(ks []string) string {
  for _, k := range ks {           
    if v := os.Getenv(k); v != "" {
      return v                     
    }                              
  }                                
  return ""                        
}                                  
```